File: games/lynrummy/python/budget_sweep.py
"""
budget_sweep.py — replay every captured projection at varying
max_states budgets. Reports, per budget, how many projections
still find a plan vs how many are sacrificed.

The OPTIMIZE_PYTHON question: can we cut the BFS state budget
without losing many plans? If the answer is "lower it 10x and
98% still find their plan," we have a free perf win at the
cost of perfect agent play on the rare hard case.

Usage:
    python3 budget_sweep.py /tmp/perf_snapshots.jsonl
"""

import argparse
import json
import sys
import time

sys.path.insert(0, ".")
import bfs
from cards import classify


BUDGETS = [200000, 50000, 20000, 10000, 5000, 2000, 1000]


def _build_initial(board, extra_stacks):
    augmented = list(board) + list(extra_stacks)
    helper = [s for s in augmented if classify(s) != "other"]
    trouble = [s for s in augmented if classify(s) == "other"]
    return (helper, trouble, [], [])


def _run_projection(initial, max_states):
    t0 = time.time()
    plan = bfs.solve_state_with_descs(
        initial, max_trouble_outer=10,
        max_states=max_states)
    wall = time.time() - t0
    return plan is not None, wall


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("snapshots")
    args = ap.parse_args()

    with open(args.snapshots) as f:
        snaps = [json.loads(l) for l in f if l.strip()]
    snaps = [s for s in snaps if s["total_wall"] < 30]

    # Reconstruct each (initial, was_plan_found, captured_wall)
    # tuple from the captured projections.
    cases = []
    for snap in snaps:
        board = [[tuple(c) for c in s] for s in snap["board"]]
        for proj in snap["projections"]:
            extra = [list(map(tuple, proj["cards"]))]
            initial = _build_initial(board, extra)
            cases.append({
                "initial": initial,
                "kind": proj["kind"],
                "captured_found": proj["found_plan"],
                "captured_wall": proj["wall"],
            })

    captured_found = sum(1 for c in cases if c["captured_found"])
    print(f"Replaying {len(cases)} projections at varying budgets.")
    print(f"Capture baseline: {captured_found}/{len(cases)} "
          f"found a plan.\n")
    print(f"{'budget':>8} {'found':>8} {'lost_vs_baseline':>18} "
          f"{'total_wall':>12}")
    print("-" * 55)

    for budget in BUDGETS:
        found_now = 0
        total_wall = 0.0
        lost = 0
        for case in cases:
            ok, wall = _run_projection(case["initial"], budget)
            total_wall += wall
            if ok:
                found_now += 1
            elif case["captured_found"]:
                lost += 1
        print(f"{budget:>8} {found_now:>8} {lost:>18} "
              f"{total_wall:>10.2f}s")


if __name__ == "__main__":
    main()

File: games/lynrummy/ts/bench/budget_sweep.ts
// budget_sweep.ts — replay every captured projection at varying
// max_states budgets. Reports, per budget, how many projections still
// find a plan vs how many are sacrificed.
//
// TS port of python/budget_sweep.py.
//
// The OPTIMIZE question: can we cut the BFS state budget without
// losing many plans? If "lower it 10× and 98% still find their plan,"
// that's a free perf win at the cost of perfect agent play on the
// rare hard case.
//
// Usage:
//   node bench/budget_sweep.ts /tmp/perf_snapshots.jsonl

import * as fs from "node:fs";

import { type Card, type Rank, type Suit, type Deck } from "../core/card.ts";
import { solveBoard } from "../bfs/index.ts";

const BUDGETS = [200000, 50000, 20000, 10000, 5000, 2000, 1000];

interface CapturedProjection {
  kind: string;
  cards: number[][];
  wall: number;
  found_plan: boolean;
}
interface CapturedSnapshot {
  hand: number[][];
  board: number[][][];
  projections: CapturedProjection[];
  total_wall: number;
  found_play: boolean;
}

function asCard(c: number[]): Card {
  return { rank: c[0]! as Rank, suit: c[1]! as Suit, deck: c[2]! as Deck };
}

function augmentedBoard(
  board: readonly (readonly Card[])[],
  extraStacks: readonly (readonly Card[])[],
): readonly (readonly Card[])[] {
  return [...board, ...extraStacks];
}

function runProjection(
  board: readonly (readonly Card[])[],
  maxStates: number,
): { found: boolean; wall: number } {
  const t0 = performance.now();
  const result = solveBoard(board, { maxTroubleOuter: 10, maxStates });
  const wall = (performance.now() - t0) / 1000;
  return { found: result !== null, wall };
}

function main(): void {
  const argv = process.argv.slice(2);
  const path = argv[0];
  if (!path) {
    process.stderr.write("usage: node bench/budget_sweep.ts <snapshots.jsonl>\n");
    process.exit(2);
    return;
  }

  const snaps: CapturedSnapshot[] = [];
  const text = fs.readFileSync(path, "utf8");
  for (const raw of text.split("\n")) {
    const line = raw.trim();
    if (!line) continue;
    const s = JSON.parse(line) as CapturedSnapshot;
    if (s.total_wall < 30) snaps.push(s);
  }

  interface Case {
    board: readonly (readonly Card[])[];
    kind: string;
    capturedFound: boolean;
    capturedWall: number;
  }

  const cases: Case[] = [];
  for (const snap of snaps) {
    const board = snap.board.map(s => s.map(asCard));
    for (const proj of snap.projections) {
      const extra = [proj.cards.map(asCard)];
      cases.push({
        board: augmentedBoard(board, extra),
        kind: proj.kind,
        capturedFound: proj.found_plan,
        capturedWall: proj.wall,
      });
    }
  }

  const capturedFound = cases.filter(c => c.capturedFound).length;
  console.log(`Replaying ${cases.length} projections at varying budgets.`);
  console.log(`Capture baseline: ${capturedFound}/${cases.length} found a plan.\n`);
  console.log(`${"budget".padStart(8)} ${"found".padStart(8)} ${"lost_vs_baseline".padStart(18)} ${"total_wall".padStart(12)}`);
  console.log("-".repeat(55));

  for (const budget of BUDGETS) {
    let foundNow = 0;
    let totalWall = 0.0;
    let lost = 0;
    for (const c of cases) {
      const { found, wall } = runProjection(c.board, budget);
      totalWall += wall;
      if (found) foundNow++;
      else if (c.capturedFound) lost++;
    }
    console.log(
      `${String(budget).padStart(8)} ${String(foundNow).padStart(8)} ${String(lost).padStart(18)} ${(totalWall.toFixed(2) + "s").padStart(12)}`,
    );
  }
}

main();
